### DISCUSSION TITLE
External call to API of module created by Module Creator

### DISCUSSION THREAD

**Discussion by avazon** [oqtane/oqtane.framework#2226] [Q&A] [answered]
Is it possible to have an external service call the API of a module created by the Module creator?
I tried to create a module with the module creator with default values. I can see the API via the swagger interface but cannot call GET, it returns error message "forbidden".
I have tried to call "login" on UserController with Postman to try to authenticate but even if I change login to "allowanonymous" it returns 403 forbidden. I suspect the UserController is not meant to be called from external service and the authentication works in a way I do not understand.

I would like to make a request with username/password and be authenticated (maybe to get a JWT token) and then be able to call the API's I create in my modules.

**Comment by sbwalker** [comment, marked as answer]
The default template in the Module Creator is designed for creating modules which integrate with the Oqtane framework. It scaffolds both the frontend UI and backend API code required and includes the authorization attributes to ensure the module is secure. If you are wanting to create an API Controller which has no frontend UI and is intended to be called by a remote client then you would be better off creating it from scratch using Visual Studio and not using the Module Creator (as you are not creating a module).

In regards to your question about the UserController, all public APIs in Oqtane are based on a backend-for-frontend (BFF) architecture which is the current recommended approach for SPA applications. A BFF architecture uses cookie authentication. So if you want to call a public API in Oqtane, you need to first login as a valid user which will create an auth cookie containing your permissions. Once you have an auth cookie you can interact with the API via Swagger, etc...

If you need to call the Oqtane public APIs from a remote client application such as Postman, there is also the ability to get a long lived JWT token by logging in to the web client as an administrator or host user and navigating to User Management in the Admin Dashboard. Choose the Settings tab and scroll down to the Token section and generate a bearer token. This is explained in this video https://www.youtube.com/watch?v=Yvp_PLvJqqk

**Reply by avazon** [reply]
Thank you for your reply. Regarding the API created by the module creator I would like to use it both from external applications and also from the frontend UI. I would use Oqtane for administer users, logging etc and in the Custom Module created show information in the frontend UI but also call the API from external applications.
I will follow your suggestion to use long lived JWT tokens for external API calls, thank you.

**Reply by avazon** [reply]
I generated a JWT token and can access the API's of Oqtane like Log, Alias, Tenant with the generated JWT token. But if I try to access an API of a custom module like the Blog module or my generated Custom Module I get 403 Forbidden. In what way is the API from a custom module different from an internal Oqtane API regarding authentication?

**Reply by sbwalker** [reply]
Security is resolved the same way for framework APIs as it is for module APIs. If you look at the methods in the HtmlTextController, BlogController, or your own custom Controller you will see that a 403 Forbidden error will be thrown if a user does not have appropriate permissions or the parameters passed to the method are invalid. In your case I am guessing you are not passing all of the required parameters when you make the API call. 

These modules all leverage the Oqtane authorization policies which allow an Administrator to dynamically specify the users/roles who have View and Edit permissions to the module. If you look at the corresponding Service classes in the Client projects, you will see that API calls use a utility method CreateAuthorizationPolicyUrl() to format the Url for the API call by including a querystring parameter "authmoduleid=##" - and in the Server Controller it uses a combination of Authorize(Policy = PolicyNames.ViewModule)] and AuthEntityId() to ensure the user is authorized to access the specific resource.

**Reply by avazon** [reply]
Thank you, you were correct. If I added ?moduleid=63&authmoduleid=63 to the query string I could access my custom module API in module #63 ( I used Fiddler to examine the traffic to find out the missing parameters)
